plugins/i3s/io/EsriReader.cpp

```diff
@@ -0,0 +1,438 @@
+                data.dimType = it->second;
+
+            if (data.dimType != Dimension::Type::None)
+            {
+                layout->registerDim(m_dimensions.at(readName));
+                m_dimMap[m_dimensions.at(readName)] = data;
+            }
+            else
+            {
+                log()->get(LogLevel::Warning) <<
+                    "Could not find Dimension type for " <<
+                    readName << std::endl;
+            }
+        }
+        else if (attributes[i].isMember("attributeValues"))
+        {
+
+            std::string s =
+                attributes[i]["attributeValues"]["valueType"].asString();
+            const pdal::Dimension::Id id = layout->registerOrAssignDim(
+                    readName, pdal::Dimension::type(s));
+            if(data.dimType != Dimension::Type::None)
+                m_dimMap[id] = data;
+        }
+        else
+        {
+            const pdal::Dimension::Id id = layout->
+                registerOrAssignDim(readName, Dimension::Type::Double);
+            m_dimMap[id] = data;
+        }
+    }
+}
+
+void EsriReader::ready(PointTableRef)
+{
+    Json::Value spatialJson = m_info["spatialReference"];
+    std::string spatialStr = "EPSG:" + spatialJson["wkid"].asString();
+    m_nativeSrs = SpatialReference(spatialStr);
+    setSpatialReference(m_nativeSrs);
+
+    m_ecefSrs = SpatialReference("EPSG:4978");
+
+    m_ecefRef = OSRNewSpatialReference(m_ecefSrs.getWKT().c_str());
+    m_nativeRef = OSRNewSpatialReference(m_nativeSrs.getWKT().c_str());
+
+    m_toEcefTransform = OCTNewCoordinateTransformation(m_nativeRef,
+            m_ecefRef);
+
+    m_toNativeTransform = OCTNewCoordinateTransformation(m_ecefRef,
+            m_nativeRef);
+}
+
+
+point_count_t EsriReader::read(PointViewPtr view, point_count_t count)
+{
+    /*
+    -3Dscenelayerinfo: URL Pattern <scene-server-url/layers/<layer-id>
+    -node index document: <layer-url >/nodepages/<iterative node page id>
+    -shared resources: <node-url>/shared/
+    -feature data: <node-url>/features/<feature-data-bundle-id>
+    -geometry data: <node-url>/geometries/<geometry-data-bundle-id>
+    -texture data: <node-url>/textures/<texture-data-bundle-id>
+    */
+
+    //Build the node list that will tell us which nodes overlap with bounds
+    std::vector<int> nodes;
+    buildNodeList(nodes, 0);
+
+    //Create view with overlapping leaf nodes
+    //Will create a thread pool on the createview class and iterate
+    //through the node list for the nodes to be pulled.
+    log()->get(LogLevel::Debug) << "Fetching binaries" << std::endl;
+    Pool p(m_args.threads);
+    for (std::size_t i = 0; i < nodes.size(); i++)
+    {
+        log()->get(LogLevel::Debug) << "\r" << i << "/" << nodes.size();
+        std::string localUrl;
+
+        localUrl = m_filename + "/nodes/" + std::to_string(nodes[i]);
+
+        p.add([localUrl, this, &view]()
+        {
+            createView(localUrl, view);
```

Comment:
Not thread-safe - pass the `PointView&` by reference.

Reply:
It's a shaued pointer. Why is this bad?

Reply:
The `shared_ptr` is passed by value (see prototype for `createView`), so each of these calls is reference-counting, which needs to be locked.

Reply:
Actually from the spec I'm not sure about this.

"If multiple threads of execution access the same shared_ptr without synchronization and any of those accesses uses a non-const member function of shared_ptr then a data race will occur".  The copy constructor is not a non-const member so maybe it's ok...